### PR TITLE
MVP with random-allocator

### PR DESCRIPTION
**Pull request by mdavezac**
- the random allocator can deal with the case where there are too many cars compared to the number of vacancies
- a python notebook (in docs/source/hotwo) demonstrates how to run the allocator with random fleet and infrastructure and gives some small analysis.
- the notebook is executed and rendered in the documentation as a howto

### CODE REVIEW

**Comment by mdavezac** [comment]
@toytun, @DT16, @cc-a, this is ready for review.

The notebook of interest is in docs/source/howto and can be started with:

```bash
jupyter notebook docs/source/howto
```

File: src/evosim/allocators.py [hunk 1]
```diff
@@ -89,7 +96,75 @@ def random_allocator(
...
+
+def random_overbooking(
+    fleet: pd.DataFrame,
+    charging_posts: pd.DataFrame,
+    method: Callable[[pd.DataFrame, pd.DataFrame], pd.DataFrame],
+    seed: Optional[Union[int, np.random.Generator]] = None,
+):
+    """Random allocation for overbooked infrastructure.
+
+    The infrastructure is overbooked when there are more electric vehicles the spare
```
Comment (cc-a) [inline src/evosim/allocators.py]:
```suggestion
    The infrastructure is overbooked when there are more electric vehicles than spare
```

File: tests/test_allocators.py [hunk 2]
```diff
@@ -1,17 +1,28 @@
...
+    evs.loc[evs_indices[: len(cps_indices)], "socket"] = cps.socket.loc[
+        cps_indices
+    ].to_numpy()
+
     matcher = matchers.factory("socket_compatibility")
-    with raises(NotImplementedError):
-        random_allocator(evs, cps, matcher, seed=rng)
+    result = random_allocator(evs, cps, matcher, seed=rng)
+    new_assignements = result.groupby("allocation").allocation.count()
+    assert (new_assignements + cps.occupancy == cps.capacity).all()
```
Comment (cc-a) [inline tests/test_allocators.py]: assignement again

**Review by cc-a — APPROVED** [review]
Looks very nice. Would suggest adding test execution of the notebook during CI to at least catch any obvious errors.

**Comment by mdavezac** [comment]
@cc-a  About the CI, the jupyter notebooks are regenerated when building the documentation. So that is already taken care of.

**Pull request merged**